otherlibs/stdlib_beta/int.ml

```diff
@@ -0,0 +1,319 @@
+      logical shift: zeroes are inserted in the vacated bits regardless
+      of the sign of [x]. The result is unspecified if [n < 0] or
+      [n >=]{!int_size}. *)
+  val shift_right_logical : t -> int -> t
+
+  (** {1:preds Predicates and comparisons} *)
+
+  (** [equal x y] is [true] if and only if [x = y]. *)
+  val equal : t -> t -> bool
+
+  (** [compare x y] is {!Stdlib.compare}[ x y] but more efficient. *)
+  val compare : t -> t -> int
+
+  (** Same as {!compare}, except that arguments are interpreted as {e unsigned} integers. *)
+  val unsigned_compare : t -> t -> int
+
+  (** Return the lesser of the two arguments. *)
+  val min : t -> t -> t
+
+  (** Return the greater of the two arguments. *)
+  val max : t -> t -> t
+
+  (** {1:convert Converting} *)
+
+  (** [to_int x] is [x] as an {!int}. If [int_size > Sys.int_size], the topmost
+      bits will be lost in the conversion *)
+  val to_int : t -> int
+
+  (** [of_int x] truncates the representation of [x] to fit in {!t}. *)
+  val of_int : int -> t
+
+  (** Same as {!to_int}, but interprets the argument as an {e unsigned} integer. *)
+  val unsigned_to_int : t -> int
+
+  (** [to_float x] is [x] as a floating point number. *)
+  val to_float : t -> float
+
+  (** [of_float x] truncates [x] to an integer. The result is
+      unspecified if the argument is [nan] or falls outside the range of
+      representable integers. *)
+  val of_float : float -> t
+
+  (** [to_string x] is the written representation of [x] in decimal. *)
+  val to_string : t -> string
+
+  (** A seeded hash function for ints, with the same output value as
+      {!Hashtbl.seeded_hash}. This function allows this module to be passed as
+      argument to the functor {!Hashtbl.MakeSeeded}. *)
+  val seeded_hash : int -> t -> int
+
+  (** An unseeded hash function for ints, with the same output value as
+      {!Hashtbl.hash}. This function allows this module to be passed as argument
+      to the functor {!Hashtbl.Make}. *)
+  val hash : t -> int
+end
+
+module Narrow
+    (Container : S) (Spec : sig
+      type t
+
+      val int_size : int
+
+      val inject : t -> Container.t
+
+      val unchecked_project : Container.t -> t
+    end) : S with type t := Spec.t = struct
+  include Spec
+
+  let () = assert (0 < int_size && int_size <= Container.int_size)
+
+  let unused_bits = Container.int_size - int_size
+
+  let[@inline] sign_extend i =
+    unchecked_project
+      (Container.shift_right (Container.shift_left i unused_bits) unused_bits)
+
+  let[@inline] project i =
+    let t = sign_extend i in
+    if Container.equal i (inject t) then Some t else None
+
+  let[@inline] zero_extend t =
+    Container.shift_right_logical
+      (Container.shift_left (inject t) unused_bits)
+      unused_bits
+
+  let zero = sign_extend Container.zero
+
+  let one = sign_extend Container.one
+
+  let minus_one = sign_extend Container.minus_one
+
+  let[@inline available] neg x = sign_extend (Container.neg (inject x))
```

Comment:
I think these should just be marked "inline" - we proactively want them to be inlined, not just if the heuristic chooses it.  But probably you had some reason in mind I'm missing?

Reply:
I was copying what int32.ml had, still inline always?

Reply:
we'll ask mark and can change later, merging as-is